pkg/service/service.go

```diff
@@ -237,10 +235,11 @@ func (s *Service) UpsertService(
 
 	backendsCopy := []lb.Backend{}
 	for _, b := range backends {
-		// Services with trafficPolicy=Local may only use node-local backends.
-		// We implement this by filtering out all backend IPs which are not a
-		// local endpoint.
-		if onlyLocalBackends && len(b.NodeName) > 0 && b.NodeName != nodeTypes.GetName() {
+		// Services with trafficPolicy=Local may only use node-local backends
+		// for external scope. We implement this by filtering out all backend
+		// IPs which are not a local endpoint.
+		if onlyLocalBackends && frontend.Scope == lb.ScopeExternal &&
```

Comment:
I think we should merge the `&& frontend.Scope == lb.ScopeExternal` condition into the `onlyLocalBackends` flag.

Or, alternatively, add `&& frontend.Scope == lb.ScopeExternal` check to the if-statement here:
https://github.com/cilium/cilium/blob/d34558669add991d90e2f346476b378413b9e205/pkg/service/service.go#L265-L268

Otherwise we will add the backends of the internal frontend to the `HealthCheckNodePort` server, which is only supposed to keep count of backends for the external fronted.

Reply:
Good point, yes, thanks for catching!